dripline/extensions/ethernet_thermo_fisher_service.py

```diff
@@ -0,0 +1,172 @@
+            commands = [commands]
+        self.alock.acquire()
+
+        try:
+            data = self._send_commands(commands)
+        except Exception as err:
+            logger.critical(str(err))
+            try:
+                data = self._send_commands(commands)
+                logger.critical("Query successful after ethernet connection recovered")
+            except socket.error as err: # simply trying to make it possible to catch the error below
+                logger.warning(f"socket.error <{err}> received, attempting reconnect")
+                logger.critical("Ethernet reconnect failed, dead socket")
+                raise ThrowReply('resource_error_connection', "Broken ethernet socket")
+            except Exception as err: ##TODO handle all exceptions, that seems questionable
+                logger.critical("Query failed after successful ethernet socket reconnect")
+                raise ThrowReply('resource_error_no_response', str(err))
+        finally:
+            self.alock.release()
+        to_return = ';'.join(data)
+        logger.debug(f"should return:\n{to_return}")
+        return to_return
+
+    def calculate_checksum(self, command_bytes):
+        """
+        Calculate the checksum by:
+        1. Summing the bytes.
+        2. Taking the lower 8 bits (1-byte sum).
+        3. Performing bitwise XOR with 0xFF.
+        """
+        byte_sum = sum(command_bytes) & 0xFF  # Sum and reduce to 1 byte
+        checksum = byte_sum ^ 0xFF            # Bitwise inversion
+        return checksum
+    
+    def check_checksum(self, cmd):
+        # calculate checksum of response except lead_char and checksum and check if match checksum
+        return self.calculate_checksum( cmd[1:-1] ) == cmd[-1]
+
+
+    def _assemble_cmd(self, cmd_in):
+        base_cmd = cmd_in[:2]
+        data = cmd_in[2:]
+        nDataBytes = len(hexstr_to_bytes(data))
+        command = hexstr_to_bytes(base_cmd) + hexstr_to_bytes(int_to_hexstr(nDataBytes)) + hexstr_to_bytes(data)
+        logger.debug(f"Assemblind cmd")
+        command = self.msb + self.lsb + command
+        logger.debug(f"Command is {command} or tpye {type(command)}")
+        cs = self.calculate_checksum(command).to_bytes(1, byteorder='big')
+        logger.debug(f"Checksum is {cs} of type {type(cs)}")
+        command = self.lead_char + command + cs
+        return command
+
+    def _send_commands(self, commands):
+        '''
+        Take a list of commands, send to instrument and receive responses, do any necessary formatting.
+
+        commands (list||None): list of command(s) to send to the instrument following (re)connection to the instrument, still must return a reply!
+                             : if impossible, set as None to skip
+        '''
+        all_data=[]
+
+        for command in commands:
+            cmd = self._assemble_cmd(command)
+            issue = True
+            try:
+                self.socket.close()
```

Comment:
is it necessary to flush the socket connection every time you want to communicate?

Reply:
May be ;) . The chiller seems to drop connection every 10 minutes. So there are "frequent" reconnects. Also a open connection blocks the serial port and does not allow to speak to the device in another way. Maybe thats wanted but at least while testing I had several times the issue that something still was connected and I could not connect to the device. 

Thats more a general question I have and I looked through several forums: What is the advantage of keeping a serial connection open vs reconnecting at each communiction. In general it sounds the prefered way is to keep the serial connection open all time but I do not see the reason for that. Is it that its blocking other connections to the device?